mix/MixClient.cpp

```diff
@@ -140,92 +196,164 @@ bytes MixClient::call(Secret _secret, u256 _value, Address _dest, bytes const& _
 	}
 	Transaction t(_value, _gasPrice, _gas, _dest, _data, n, _secret);
 	bytes rlp = t.rlp();
-	WriteGuard lw(x_state); //TODO: lock is required only for last executoin state
-	executeTransaction(&rlp, temp);
-	return m_lastExecutionResult.returnValue;
+	WriteGuard lw(x_state); //TODO: lock is required only for last execution state
+	executeTransaction(t, temp);
+	return m_blocks.back().transactions.back().returnValue;
 }
 
 u256 MixClient::balanceAt(Address _a, int _block) const
 {
-	validateBlock(_block);
 	ReadGuard l(x_state);
-	return m_state.balance(_a);
+	return asOf(_block).balance(_a);
 }
 
 u256 MixClient::countAt(Address _a, int _block) const
 {
-	validateBlock(_block);
 	ReadGuard l(x_state);
-	return m_state.transactionsFrom(_a);
+	return asOf(_block).transactionsFrom(_a);
 }
 
 u256 MixClient::stateAt(Address _a, u256 _l, int _block) const
 {
-	validateBlock(_block);
 	ReadGuard l(x_state);
-	return m_state.storage(_a, _l);
+	return asOf(_block).storage(_a, _l);
 }
 
 bytes MixClient::codeAt(Address _a, int _block) const
 {
-	validateBlock(_block);
 	ReadGuard l(x_state);
-	return m_state.code(_a);
+	return asOf(_block).code(_a);
 }
 
 std::map<u256, u256> MixClient::storageAt(Address _a, int _block) const
 {
-	validateBlock(_block);
 	ReadGuard l(x_state);
-	return m_state.storage(_a);
+	return asOf(_block).storage(_a);
 }
 
 eth::LocalisedLogEntries MixClient::logs(unsigned _watchId) const
 {
-	(void)_watchId;
-	return LocalisedLogEntries();
+	Guard l(m_filterLock);
+	h256 h = m_watches.at(_watchId).id;
+	auto filterIter = m_filters.find(h);
+	if (filterIter != m_filters.end())
+		return logs(filterIter->second.filter);
+	return eth::LocalisedLogEntries();
 }
 
-eth::LocalisedLogEntries MixClient::logs(eth::LogFilter const& _filter) const
```

Comment:
The pyramid below is visually quite appealing, so I do not dare to destroy it by a comment, but I am not sure if it is easily understandable. Please expand the names of some of the variables (it took me quite some minutes to check where `s` is actually used and that might be the number of matching logs that are to be skipped....).
It perhaps also might be useful to not introduce `m` at all. I also like to compare running indices against sizes using `<` or `<=` instead of `!=` to indicate which one is the running index and which is the size, but that is probably a matter of taste.